tests/Crypto/cryptotest.cpp

```diff
@@ -0,0 +1,158 @@
+    m_testPluginCryptoStorage = findPluginName("TEST_PLUGIN_CRYPTO_STORAGE", CryptoManager::DefaultCryptoStoragePluginName + QLatin1String(".test"));
+    m_testPluginStorage = findPluginName("TEST_PLUGIN_STORAGE", Sailfish::Secrets::SecretManager::DefaultStoragePluginName + QLatin1String(".test"));
+    m_testPluginEncryptedStorage = findPluginName("TEST_PLUGIN_ENCRYPTED_STORAGE", Sailfish::Secrets::SecretManager::DefaultEncryptedStoragePluginName + QLatin1String(".test"));
+    m_testPluginAuth = findPluginName("TEST_PLUGIN_AUTH", Sailfish::Secrets::SecretManager::DefaultAuthenticationPluginName + QLatin1String(".test"));
+    m_testPluginInAppAuth = findPluginName("TEST_PLUGIN_IN_APP_AUTH", Sailfish::Secrets::SecretManager::InAppAuthenticationPluginName + QLatin1String(".test"));
+    m_testPluginEncryption = findPluginName("TEST_PLUGIN_ENCRYPTION", Sailfish::Secrets::SecretManager::DefaultEncryptionPluginName + QLatin1String(".test"));
+    m_testPluginUsbToken = findPluginName("TEST_PLUGIN_USB_TOKEN", QStringLiteral("org.sailfishos.secrets.plugin.cryptostorage.exampleusbtoken.test"));
+}
+
+QByteArray CryptoTest::createRandomTestData(int size)
+{
+    QFile file("/dev/urandom");
+    if (file.open(QIODevice::ReadOnly)) {
+        QByteArray result = file.read(size);
+        file.close();
+        return result;
+    }
+    qWarning() << "Cannot open" << file.fileName();
+    return QByteArray();
+}
+
+QByteArray CryptoTest::generateInitializationVector(Sailfish::Crypto::CryptoManager::Algorithm algorithm,
+                                                    Sailfish::Crypto::CryptoManager::BlockMode blockMode)
+{
+    if (algorithm != CryptoManager::AlgorithmAes || blockMode == CryptoManager::BlockModeEcb) {
+        return QByteArray();
+    }
+    switch (blockMode) {
+        case CryptoManager::BlockModeGcm:
+            return createRandomTestData(12);
+        case CryptoManager::BlockModeCcm:
+            return createRandomTestData(7);
+    default:
+        break;
+    }
+    return createRandomTestData(16);
+}
+
+bool CryptoTest::allCharactersAreNull(const QString &s)
+{
+    for (auto c : s) {
+        if (c != '\0') {
+            return false;
+        }
+    }
+    return true;
+}
+
+bool CryptoTest::needsAuthentication(Sailfish::Crypto::CryptoManager::Algorithm algorithm,
+                                     Sailfish::Crypto::CryptoManager::BlockMode blockMode)
+{
+    if (m_testPluginCrypto == CryptoManager::DefaultCryptoPluginName + QLatin1String(".test")) {
+        return algorithm == Sailfish::Crypto::CryptoManager::AlgorithmAes
+                && (blockMode == Sailfish::Crypto::CryptoManager::BlockModeGcm
+                    || blockMode == Sailfish::Crypto::CryptoManager::BlockModeCcm);
+    }
+    qWarning() << "Test is not using the default test plugin, assuming authentication data is not required";
```

Comment:
Why do we only test the authentication data flows with the default plugin?

Reply:
I remember discussing that plugins would have their own authentication requirement configurations for custom algorithms and block modes, which we can't know. What do you think is a good way to handle that?

Reply:
I think that it should be denoted by a QTest data column value.  The unit test itself should then do something like:

if (operationXSupported) {
    // ... test operation X ...
    QCOMPARE(resultValue, expectedOperationXResultValue);
}

Reply:
Okay, yep the plugin test configuration could specify the algorithms and block modes that require authentication so that we can configure this accordingly.